db-init.sql

```diff
@@ -0,0 +1,166 @@
+	-- all the different types a person can be
+
+CREATE TABLE PersonCategories (
+	personCategoryId SERIAL PRIMARY KEY NOT NULL,
+	personCategoryName VARCHAR(255) NOT NULL,
+);
+
+	CREATE UNIQUE INDEX personCategoryName_lower ON PersonCategories(lower(personCategoryName));
+
+	INSERT INTO PersonCategories VALUES (0, 'farmers');
+	INSERT INTO PersonCategories VALUES (1, 'traders');
+	INSERT INTO PersonCategories VALUES (2, 'admins');
+	INSERT INTO PersonCategories VALUES (3, 'monitor');
+
+-- PersonAttributeTypes
+	--  all the different types of attributes a person can have
+
+CREATE TABLE PersonAttributeTypes (
+	attrId SERIAL PRIMARY KEY NOT NULL,
+	attrName VARCHAR(255) NOT NULL
+);
+
+	CREATE UNIQUE INDEX personAttrName_lower ON PersonAttributeTypes(lower(attrName));
+
+	INSERT INTO PersonAttributeTypes VALUES (0, 'username');
+	INSERT INTO PersonAttributeTypes VALUES (1, 'passwordHash');
+	INSERT INTO PersonAttributeTypes VALUES (2, 'paymentFrequency');
+	INSERT INTO PersonAttributeTypes VALUES (3, 'notes');
+
+-- PersonCategoryAttributes
+	-- what types of attributes a category of people have
+
+CREATE TABLE PersonCategoryAttributes (
+	personCategoryId SERIAL REFERENCES PersonCategories(personCategoryId) NOT NULL,
+	attrId SERIAL REFERENCES PersonAttributeTypes(attrId) NOT NULL
+);
+
+	-- farmer
+	INSERT INTO PersonCategoryAttributes VALUES (0, 2); -- paymentFrequency
+	INSERT INTO PersonCategoryAttributes VALUES (0, 3); -- notes
+	-- trader
+	INSERT INTO PersonCategoryAttributes VALUES (1, 0); -- username
+	INSERT INTO PersonCategoryAttributes VALUES (1, 1); -- password
+	-- admin
+	INSERT INTO PersonCategoryAttributes VALUES (2, 0); -- username
+	INSERT INTO PersonCategoryAttributes VALUES (2, 1); -- password
+	-- admin
+	INSERT INTO PersonCategoryAttributes VALUES (3, 0); -- username
+	INSERT INTO PersonCategoryAttributes VALUES (3, 1); -- password
+
+-- Person
+
+CREATE TABLE Persons (
+	personUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+	firstName VARCHAR(255),
+	middleName VARCHAR(255),
+	lastName VARCHAR(255),
+	phoneNumber VARCHAR(20),
+	phoneArea VARCHAR(20),
+	phoneCountry VARCHAR(20),
+	companyName VARCHAR(255),
+	personCategoryId SERIAL REFERENCES PersonCategories(personCategoryId)
+);
+
+-- PersonAttributes
+	-- the values of the attributes that a specific person has
+
+CREATE TABLE PersonAttributes (
+	personUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	attrId SERIAL REFERENCES PersonAttributeTypes(attrId) NOT NULL,
+	attrValue VARCHAR(255)
+);
+
+-- MoneyTransactions
+
+CREATE TABLE MoneyTransactions (
+	moneyTransactionUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+	datetime TIMESTAMP WITH TIME ZONE NOT NULL,
+	toPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	fromPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	amount MONEY NOT NULL
+);
+
+-- ProductTypes
+	-- all the different kinds of products
+
+CREATE TABLE ProductTypes (
+	productTypeId SERIAL PRIMARY KEY NOT NULL,
```

Comment:
would this support having milk in kg as well? it would just be a different id right?

Reply:
No it's not supported due to line 94. That case you described is something I want to avoid because those make it hard to organize data and query it in the database. Everything should be converted to a unified format before being put in the database.

```sql
CREATE UNIQUE INDEX productNames_lower ON ProductTypes(lower(productName));
```